### PR TITLE
adding autoware_auto_perception_rviz_plugin

### PR DESCRIPTION
**Pull request by taichiH**
## Description

Adding autoware_auto_perception_rviz_plugin to visualize `DetectedObjects` , `TrackedObjects` and `PredictedObjects` type messages.

The code is based on the following two packages. 
- https://gitlab.com/autowarefoundation/autoware.auto/AutowareAuto/-/tree/master/src/tools/visualization/autoware_rviz_plugins
- https://github.com/tier4/AutowareArchitectureProposal.iv/tree/use-autoware-auto-msgs/perception/util/visualizer/dynamic_object_visualization

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Read [pull request guidelines][pull-request-guidelines]
- [x] Code follows [coding guidelines][coding-guidelines]
- [x] Assign PR to reviewer

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] Code follows [coding guidelines][coding-guidelines]
- [ ] (Optional) Unit tests have been written for new behavior
- [ ] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets
- [ ] Write [release notes][release-notes]

## CI Checks

- **Build and test for PR / build-and-test-pr**: Required to pass before the merge.
- **Build and test for PR / clang-tidy-pr**: NOT required to pass before the merge. It is up to the reviewer(s). Found false positives? See the [guidelines][clang-tidy-guidelines].
- **Check spelling**: NOT required to pass before the merge. It is up to the reviewer(s). See [here][spell-check-dict] if you want to add some words to the spell check dictionary.

[clang-tidy-guidelines]: https://tier4.github.io/autoware.proj/tree/main/developer_guide/ClangTidyGuideline/
[coding-guidelines]: https://tier4.atlassian.net/wiki/spaces/AIP/pages/1194394777/T4
[pull-request-guidelines]: https://tier4.github.io/autoware.proj/tree/main/developer_guide/PullRequestGuideline/
[release-notes]: https://tier4.atlassian.net/wiki/spaces/AIP/pages/563774416
[spell-check-dict]: https://github.com/tier4/autoware-spell-check-dict#how-to-contribute

### CODE REVIEW

**Comment by taichiH** [comment]
This is visualization result.
https://drive.google.com/file/d/1WnozQz7WGr4jr8f1BqvdwIGVyT19N9EH/view?usp=sharing

**Comment by taichiH** [comment]
@yukkysaito 
Should I remove oridinary dynamic_object_visualization package?

File: common/util/rviz_plugins/autoware_auto_perception_rviz_plugin/include/object_detection/object_polygon_detail.hpp [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+     {"UNKNOWN", {255, 255, 255}}},
+    {autoware_auto_perception_msgs::msg::ObjectClassification::CAR, {"CAR", {0, 255, 142}}},
+    {autoware_auto_perception_msgs::msg::ObjectClassification::BUS, {"BUS", {0, 200, 100}}},
+    {autoware_auto_perception_msgs::msg::ObjectClassification::PEDESTRIAN,
+     {"PEDESTRIAN", {220, 20, 60}}},
+    {autoware_auto_perception_msgs::msg::ObjectClassification::BICYCLE, {"CYCLIST", {119, 11, 32}}},
+    {autoware_auto_perception_msgs::msg::ObjectClassification::MOTORCYCLE,
+     {"MOTORCYCLE", {0, 0, 230}}},
+    {autoware_auto_perception_msgs::msg::ObjectClassification::TRAILER, {"TRAILER", {0, 80, 100}}},
+    {autoware_auto_perception_msgs::msg::ObjectClassification::TRUCK, {"TRUCK", {0, 0, 70}}}};
```
Comment (taichiH) [inline common/util/rviz_plugins/autoware_auto_perception_rviz_plugin/include/object_detection/object_polygon_detail.hpp:69]: @yukkysaito 
We can change the polygon color for each label.
If you have color request, please let me know.

**Comment by yukkysaito** [comment]
>Should I remove oridinary dynamic_object_visualization package?

Yes

Can you give me a sample image of what it looks like to visualize?

**Comment by taichiH** [comment]
@yukkysaito 

> > Should I remove oridinary dynamic_object_visualization package?
> > @taichiH
> > Yes
> 
> Can you give me a sample image of what it looks like to visualize?

Here is a sample screen capture.
https://drive.google.com/file/d/1WnozQz7WGr4jr8f1BqvdwIGVyT19N9EH/view?usp=sharing

**Comment by yukkysaito** [comment]
@taichiH めんどくさくなってきたので日本語ですいません
image
こちらのCARとかの中にはどういう設定項目がありますか？

**Comment by taichiH** [comment]
@yukkysaito There are properties for setting color and alpha.
Screenshot from 2021-11-10 16-41-33

**Comment by yukkysaito** [comment]
いいすねｗ

**Comment by yukkysaito** [comment]
Great

**Comment by wep21** [comment]
@taichiH Could you add icons for plugins?

File: common/util/rviz_plugins/autoware_auto_perception_rviz_plugin/include/object_detection/object_polygon_detail.hpp (hunk 1 above)
Comment (yukkysaito) [reply common/util/rviz_plugins/autoware_auto_perception_rviz_plugin/include/object_detection/object_polygon_detail.hpp:69]: I may change it later, but can you do this for now?
pedestrian : rgb(255,192,203)
vehicle : rgb(30,144,255)
motorcycle : rgb(255,204,102)

File: common/util/rviz_plugins/autoware_auto_perception_rviz_plugin/include/object_detection/object_polygon_display_base.hpp [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+protected:
+  /// \brief Convert given shape msg into a Marker
+  /// \tparam ClassificationContainerT List type with ObjectClassificationMsg
+  /// \param shape_msg Shape msg to be converted
+  /// \param centroid Centroid position of the shape in Object.header.frame_id frame
+  /// \param orientation Orientation of the shape in Object.header.frame_id frame
+  /// \param labels List of ObjectClassificationMsg objects
+  /// \return Marker ptr. Id and header will have to be set by the caller
+  template <typename ClassificationContainerT>
+  boost::optional<Marker::SharedPtr> get_shape_marker_ptr(
```
Comment (wep21) [inline common/util/rviz_plugins/autoware_auto_perception_rviz_plugin/include/object_detection/object_polygon_display_base.hpp:137]: galactic supports C++17, so `std::optional` is better.

File: common/util/rviz_plugins/autoware_auto_perception_rviz_plugin/include/object_detection/object_polygon_display_base.hpp [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+        std::forward_as_tuple(
+          QColor{color[0], color[1], color[2]}, class_property_values.alpha, &parent_property));
+    }
+    init_color_list(predicted_path_colors);
+  }
+
+  void onInitialize() override
+  {
+    RosTopicDisplay::RTDClass::onInitialize();
+    m_marker_common.initialize(this->context_, this->scene_node_);
```
Comment (wep21) [inline common/util/rviz_plugins/autoware_auto_perception_rviz_plugin/include/object_detection/object_polygon_display_base.hpp:102]:
```suggestion
    m_marker_common.initialize(this->context_, this->scene_node_);
    QString message_type = QString::fromStdString(rosidl_generator_traits::name<MsgT>());
    this->topic_property_->setMessageType(message_type);
```

File: common/util/rviz_plugins/autoware_auto_perception_rviz_plugin/include/object_detection/object_polygon_display_base.hpp [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+  using Color = std::array<float, 3U>;
+  using Marker = visualization_msgs::msg::Marker;
+  using MarkerCommon = rviz_default_plugins::displays::MarkerCommon;
+  using ObjectClassificationMsg = autoware_auto_perception_msgs::msg::ObjectClassification;
+  using RosTopicDisplay = rviz_common::RosTopicDisplay<MsgT>;
+
+  using PolygonPropertyMap =
+    std::unordered_map<ObjectClassificationMsg::_label_type, common::ColorAlphaProperty>;
+
+  explicit ObjectPolygonDisplayBase(const std::string & default_topic)
```
Comment (wep21) [inline common/util/rviz_plugins/autoware_auto_perception_rviz_plugin/include/object_detection/object_polygon_display_base.hpp:61]: [nits] Is default topic necessary?

File: common/util/rviz_plugins/autoware_auto_perception_rviz_plugin/include/object_detection/object_polygon_detail.hpp (hunk 1 above)
Comment (taichiH) [reply common/util/rviz_plugins/autoware_auto_perception_rviz_plugin/include/object_detection/object_polygon_detail.hpp]: I changed colors.

https://github.com/tier4/AutowareArchitectureProposal.iv/pull/574/commits/54c53389cc6d50ead3012eceacddf73e5117d1df
Screenshot from 2021-11-11 09-07-45

File: common/util/rviz_plugins/autoware_auto_perception_rviz_plugin/include/object_detection/object_polygon_display_base.hpp (hunk 2 above)
Comment (taichiH) [reply common/util/rviz_plugins/autoware_auto_perception_rviz_plugin/include/object_detection/object_polygon_display_base.hpp]: It was replaced to std::optional on https://github.com/tier4/AutowareArchitectureProposal.iv/pull/574/commits/fd0b931837052c8c7d570e6e7bea66f050d9609b

**Comment by taichiH** [comment]
@wep21 

> @taichiH Could you add icons for plugins?

I added icons. https://github.com/tier4/AutowareArchitectureProposal.iv/pull/574/commits/10ffda1488bd0d1d35020434db1828e202064d10

File: common/util/rviz_plugins/autoware_auto_perception_rviz_plugin/include/object_detection/object_polygon_detail.hpp (hunk 1 above)
Comment (yukkysaito) [reply common/util/rviz_plugins/autoware_auto_perception_rviz_plugin/include/object_detection/object_polygon_detail.hpp]: Can you make bicycle the same color as motorcycle?

Comment (taichiH) [reply common/util/rviz_plugins/autoware_auto_perception_rviz_plugin/include/object_detection/object_polygon_detail.hpp]: https://github.com/tier4/AutowareArchitectureProposal.iv/pull/574/commits/0abb30505e7b668f6be6dc9d6230b8984f6bdf6f

**Review by wep21 — APPROVED** [review]

**Pull request merged**